File: collector/core/registry.go
package core

import (
	"context"
	"fmt"
	"sort"
)

// Source collects a batch of messages. Implementations are constructed once
// at startup and Collect is called on every poll.
type Source interface {
	Collect(ctx context.Context) ([]Message, error)
}

// Destination receives batches of collected messages.
type Destination interface {
	Write(ctx context.Context, batch []Message) error
	Close(ctx context.Context) error
}

// SourceCtor builds a Source from its YAML `config` block.
type SourceCtor func(cfg Fields) (Source, error)

// DestinationCtor builds a Destination from its YAML `config` block.
type DestinationCtor func(cfg Fields) (Destination, error)

var (
	sourceTypes      = map[string]SourceCtor{}
	destinationTypes = map[string]DestinationCtor{}
)

// RegisterSource makes a source type available under the given name. Call it
// from an init function in the file implementing the type.
func RegisterSource(typeName string, ctor SourceCtor) {
	if _, exists := sourceTypes[typeName]; exists {
		panic(fmt.Sprintf("source type %q registered twice", typeName))
	}
	sourceTypes[typeName] = ctor
}

// RegisterDestination makes a destination type available under the given
// name. Call it from an init function in the file implementing the type.
func RegisterDestination(typeName string, ctor DestinationCtor) {
	if _, exists := destinationTypes[typeName]; exists {
		panic(fmt.Sprintf("destination type %q registered twice", typeName))
	}
	destinationTypes[typeName] = ctor
}

// NewSource instantiates a registered source type.
func NewSource(typeName string, cfg Fields) (Source, error) {
	ctor, ok := sourceTypes[typeName]
	if !ok {
		return nil, fmt.Errorf("unknown source type %q, available: %v", typeName, SourceTypes())
	}
	return ctor(cfg)
}

// NewDestination instantiates a registered destination type.
func NewDestination(typeName string, cfg Fields) (Destination, error) {
	ctor, ok := destinationTypes[typeName]
	if !ok {
		return nil, fmt.Errorf("unknown destination type %q, available: %v", typeName, DestinationTypes())
	}
	return ctor(cfg)
}

// SourceTypes lists all registered source type names.
func SourceTypes() []string {
	names := make([]string, 0, len(sourceTypes))
	for name := range sourceTypes {
		names = append(names, name)
	}
	sort.Strings(names)
	return names
}

// DestinationTypes lists all registered destination type names.
func DestinationTypes() []string {
	names := make([]string, 0, len(destinationTypes))
	for name := range destinationTypes {
		names = append(names, name)
	}
	sort.Strings(names)
	return names
}

File: collector-rs/src/registry.rs
use std::collections::BTreeMap;

use anyhow::Result;
use async_trait::async_trait;

use crate::message::Message;

/// A source collects a batch of messages. Constructed once at startup,
/// `collect` is called on every poll.
#[async_trait]
pub trait Source: Send + Sync {
    async fn collect(&self) -> Result<Vec<Message>>;
}

/// A destination receives batches of collected messages.
#[async_trait]
pub trait Destination: Send + Sync {
    async fn write(&self, batch: &[Message]) -> Result<()>;
    async fn close(&self) -> Result<()> {
        Ok(())
    }
}

/// Builds a source from its YAML `config` block.
pub type SourceFactory = fn(&serde_yaml::Value) -> Result<Box<dyn Source>>;

/// Builds a destination from its YAML `config` block.
pub type DestinationFactory = fn(&serde_yaml::Value) -> Result<Box<dyn Destination>>;

/// The registry of known source and destination types. Adding a new type is
/// one line in `builtin()` plus a module implementing the trait.
pub struct Registry {
    sources: BTreeMap<&'static str, SourceFactory>,
    destinations: BTreeMap<&'static str, DestinationFactory>,
}

impl Registry {
    /// Returns the registry of built-in types.
    pub fn builtin() -> Self {
        let mut r = Self {
            sources: BTreeMap::new(),
            destinations: BTreeMap::new(),
        };
        r.register_source("http_api", crate::sources::http_api::new);
        r.register_source("redfish", crate::sources::redfish::new);
        r.register_source("oc", crate::sources::oc::new);
        r.register_source("vcenter", crate::sources::vcenter::new);
        r.register_destination("stdout", crate::destinations::stdout::new);
        r.register_destination("http_post", crate::destinations::http_post::new);
        r.register_destination("mongodb", crate::destinations::mongodb::new);
        r
    }

    pub fn register_source(&mut self, name: &'static str, factory: SourceFactory) {
        if self.sources.insert(name, factory).is_some() {
            panic!("source type {name:?} registered twice");
        }
    }

    pub fn register_destination(&mut self, name: &'static str, factory: DestinationFactory) {
        if self.destinations.insert(name, factory).is_some() {
            panic!("destination type {name:?} registered twice");
        }
    }

    pub fn new_source(&self, type_name: &str, cfg: &serde_yaml::Value) -> Result<Box<dyn Source>> {
        match self.sources.get(type_name) {
            Some(factory) => factory(cfg),
            None => anyhow::bail!(
                "unknown source type {:?}, available: {:?}",
                type_name,
                self.source_types()
            ),
        }
    }

    pub fn new_destination(
        &self,
        type_name: &str,
        cfg: &serde_yaml::Value,
    ) -> Result<Box<dyn Destination>> {
        match self.destinations.get(type_name) {
            Some(factory) => factory(cfg),
            None => anyhow::bail!(
                "unknown destination type {:?}, available: {:?}",
                type_name,
                self.destination_types()
            ),
        }
    }

    pub fn source_types(&self) -> Vec<&'static str> {
        self.sources.keys().copied().collect()
    }

    pub fn destination_types(&self) -> Vec<&'static str> {
        self.destinations.keys().copied().collect()
    }
}
